### ISSUE TITLE
Slow Cobbler startup time

### ISSUE THREAD

**Issue by SchoolGuy** [cobbler/cobbler#2804]
### Describe the bug

Currently due to additional `serialize()` calls Cobbler is starting slow, compared to before.

### Steps to reproduce

1. Fill in a lot of items
2. Restart the demon
3. Notice slow startup times

### Expected behavior

Cobbler should not try to write data back to the disk when starting up. This should be done during changes of the data.

### Cobbler version

Custom version for the Orthos 2 project which is equal to V3.3.0

### Operating system

openSUSE Leap 15.3

### Cobbler log

Not applicable

### Screenshots

Not applicable

### Additional information

Not applicable

**Comment by SchoolGuy** [comment]
So the reason for this is that Cobbler tries to write the collections back to the disk once the `parent` or `children` do change in the in-memory object. This can be fully removed of course but that would mean that we don't have a correct state of the object represented on disk which leads to data loss ultimately when Cobbler crashes or the demon is stopped.

Solutions:

1. Accepting the slow startup (not acceptable in my eyes)
2. Regular Sync intervals (to error prone in my eyes because it produces unnecessary load)
3. Complete decoupling from in-memory and on-disk state (hard to engineer into the application but probably the best solution)
4. Enabling per-object(-tree) serialization (probably good to have anyway)
5. Checking if the content on disk would change and only write if this is the case

I would vote for a combination of 3 & 4. However solution 5 also is appealing but it is probably to complex for the small amount of data we want to write.

The status quo currently is that once data changes, all items are serialized.

**Comment by watologo1** [comment]
Thanks for the overview!

I guess this boils down to:
1. Per object serialization
2. Never serialize when data got read (getter), check if data changed when written to it (setter)

We should go for the long hanging fruits first...
If I understand things correctly, currently algorithm writes n*n objects, even on startup?
This explains why the latter objects take much longer to add than the first ones.

Obviously, data need not get synced at all at startup, the data is simply read from disk.
Not sure how complex this is and fits in current structure, but I guess the setter need to know whether it's the initial object initialization. Also currently the setter also validates against its own data in cobbler profile/system/distro edit case. This sounds related as well.
Maybe each item could have a dirty state it internally updates itself and its serialze() function would only do something if dirty == True ?
dirty must not be set at object deserialization (setters used but from origin disk data) time.
dirty could also be protected by a semaphore. This would serialize all write data and would avoid that data is only partly overridden by one thread. This would also enforce that the thread who called any setter on the object, would have to call
item.serialize() afterwards. Some kind of timeout and backtrace pointing to the code where the setter was used, but did not call save() (needs proper exception handling).
Not sure whether this is what we want, but this could help as an example how define selfmade @classproperty instead of @property and let the __set__ method for example set an internal __dirty__ = True.

I do absolutely not have time to look into this further, even I'd like to.
This could be a non-thread save implementation. Making it thread safe would need adding some kind of semaphore/lock.

This problem is rather generic (serialize python objects thread safe and after finishing wtih all getters) and maybe one finds a better solution, ideally with a tiny prove of concept code example...

Note: rename is evil -> Ideally I would try to set this "deprecated" and let people do: remove/add from outside instead.
Just an idea...

**Comment by watologo1** [comment]
Adding some people with deeper python knowledge..., this is a complex topic.
There also might be newer python3.x elements...

**Comment by SchoolGuy** [comment]
The problem we have is a chicken egg one: We want to set data to initialize the objects but since we don't have full control over the data (could be manually messed up) we don't trust it. Thus we use the properties to set the data. Because we use the setter for the initialization we then write the valid data back into the JSON which currently is being loaded. This is not a problem in 99% of the cases since the content doesn't change and we have the file(s) open anyway.

We can't deprecate `rename` because it is one of the basic options. The root problem is that Cobbler works internally with the name as a primary key, although it is not a primary key in reality. We have `uid`s which we should use for that, then a rename wouldn't be a problem at all.

**Comment by SchoolGuy** [comment]
@meaksh @agraul @vzhestkov Any ideas how to solve this chicken-egg problem?

**Comment by SchoolGuy** [comment]
@bill-broadley you might be interested in this issue. They are not identical but they might have the same root-cause.

**Comment by SchoolGuy** [comment]
@AaronBandt So to give you more context here that is not 100% visible already in this issue:

We have additionally to this issue another performance related issue (#3259). Those issues basically have the same root cause to my current investigation. Currently I have defined that the JSON database is not to be trusted and also data coming from the outside (so XML-RPC essentially) is not to be taken as valid. Since I want to validate this potentially invalid data I asked myself what the lowest common point of data passing through the application is. After some debugging I found that this are the items. Since back in 2.8.5 we just had plain objects, I decided to change that to properties (#2433). The idea behind this was to introduce validation at a single point and not duplicate the code of validation across the codebase. Also this guaranteed us that data coming from a point that I would have not thought of (for example a custom module or trigger) is still validated. Sadly it shows slowly now that while this approach guarantees that Cobbler runs much more resilient (imho) in case invalid data is submitted to the application, it means it is much slower.

So to the point of how to solve this... Well I am no genius but I am trying my best. Time is my biggest constraint and at work my priorities have shifted temporarily (I am on a team exchange), however I am still developing during weekends. So far I am still out of ideas how to change this with smaller changes without sacrificing data integrity during storage and runtime. Bigger changes are possible but they take time and this is what I lack of atm. Contributions are very welcome, I am eager to support them if they come.

TLDR: A proper solution for this is one that doesn't sacrifice data integrity but enables performance gains during startup (and thus hopefully during the rest of the runtime). How that is achieved I don't really care atm if the solution is sustainable and not hacky.

**Comment by SchoolGuy** [comment]
@AaronBandt Well there are two reasons...

1. I am paranoid because this is a tool for sysadmins. They like to play and the easy datastructure is inviting them to do so 
2. Databases like Mariadb or other ones are also performing checks during startup if I am not mistaken. The application is just not seeing this as the DB is normally already running when the application starts.

**Comment by SchoolGuy** [comment]
This is closed in favour of #3360. I consider this a duplicate.
